### ISSUE TITLE
"Reflected list" is being deprecated when there is no reflection?

### ISSUE THREAD

**Issue by mehrdadn** [numba/numba#4158]
Hi,

When I run this code:

	@njit
	def test(arr):
		return [0]
	test([0])

I get the error:

     Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'arr' of function 'test'.

However, I think this is a false positive given that there is no reflection being performed in the function. Is this behavior genuinely being deprecated, and if so, should it in fact be deprecated? It seems find to me to allow this given that the list isn't being written to at all, but I"m not sure what all the implications are.

Thank you!

**Comment by stuartarchibald** [comment]
Thanks for the report. The warning is issued based on finding an inferred type of `reflected list`, all lists passed as arguments will be this type irrespective of whether they are mutated. The reason in part for this, specific to this case, is that it is hard in general to tell if a list is being mutated (consider aliasing or a write in a branch based taken based on the value of an input argument).

The behaviour above will cease to work once `numba.typed.List` is implemented and list reflection is removed so I think the above message is valid.

**Comment by mehrdadn** [comment]
Wouldn't it be easy to detect if a list might be *potentially* mutated at least (regardless of whether branches are actually taken), and, if it is not, then allow it to be used without reflection? It seems quite reasonable for the above code to keep working on that basis...

**Comment by sklam** [comment]
Reflection happens eagerly when a list is passed as argument to a jit'ed function irregardless of whether the list is used.  But you are right that this type of non-mutating list usage should keep working.

**Comment by mehrdadn** [comment]
At jit-time it should be easy to tell that that there is no write instruction to the list in the body, in which case reflection should always be suppressed for that argument. There's no need to perform reflection all the time.

**Comment by sklam** [comment]
Only for some types.  Others like array types, the mutation is far too complicated to detect.  One worry is that a conservative detection will make it difficult for users to understand what works and what doesn't.

**Comment by sklam** [comment]
> But you are right that this type of non-mutating list usage should keep working.

After discussing with the core devs, I recalled why even non-mutating list is going to be a challenge.  

* When a Python list is used as argument to a jit'ed function, every item it holds must be scanned to determine the type.  This leads to an unexpected performance hit for the users.  If it's a list of heterogeneous types, the types will be coerced, which is the current behavior but also unexpected for users.
* Mutation is hard to detect for the general case.  For a non-nested list of immutable types (i.e. ints, floats), the only mutation is by direct mutation on the list.  That is the simple case.  (Maybe we can do something special for this case.)  For others, there are indirect mutations, such as self-mutating methods in element of the list.

The current plan is to switch to a more explicit form for handling list.  Such that users can better control the type and better understand when costly operations are performed.  

For details, please see http://numba.pydata.org/numba-doc/latest/reference/deprecation.html?highlight=deprecation#deprecation-of-reflection-for-list-and-set-types.  And, please let us know what you think.

**Comment by mehrdadn** [comment]
Thanks. Yup I read that page before coming here; it specifically talks about deprecation of *reflection* which reinforced to me the notion that non-reflected lists shouldn't be disallowed.

Re: mutation being hard to detect: again, you don't need to. The rules only need to be exactly the same as `final` in Java or `const` for variables in C/C++ -- as long as you don't even pretend to write to the the target, the code is valid. Beginners tend to grasp this pretty easily. (Although I would also debate the notion that the target of the library should be beginners who are new to programming...)

Re: indirect mutations: this does sound like potentially a genuine problem, I'm not sure I see how it arises in the first place (and I also think it shouldn't preclude arrays of primitives). Can a list of arbitrary objects be passed to nopython code? Would you mind providing an example of the situation?

Re: scanning being O(n): this is the most compelling argument to me. In terms of homogeneity here, I agree it must be a requirement (it should just be an error otherwise), but scanning genuinely can't be avoided altogether. Can I ask what the "more explicit" forms would look like in the new plan?

Thanks!

**Comment by seibert** [comment]
The pattern for list support we are aiming for (will take several steps due to the need to deprecate the existing support) will be similar to how we implemented dictionary support:

http://numba.pydata.org/numba-doc/latest/reference/pysupported.html#dict

For lists it will go something like:

* Within nopython mode, creating a list literal (`[1, 2]`) will create a typed list where the element type is known.  Numba's type inference will determine this type from the way the list is used.
* A list returning from nopython mode will be boxed into a `numba.typed.List` object which functionally behaves like a list, but uses an internal storage that has no Python objects.  These typed list objects can be passed with minimal overhead to other Numba-compiled functions, or used directly from the Python interpreter.  (Interpreter get/set item will necessarily have more overhead than a Python list because Numba will have to unbox and box elements on their way in and out.)
* Users can convert their Python list to a Numba typed list outside of nopython mode in order to take the conversion hit only once at a location of their choosing in the code, rather than having to do it on entry to every Numba-compiled function.

**Comment by mehrdadn** [comment]
Interesting, thanks. Tangent, but out of curiosity, why not just use `array.array` instead of a custom `numba.typed.List` type? It's already serving the role of being a homogeneous list that behaves like `list`.

**Comment by seibert** [comment]
The list element type in a Numba typed list will be any type that Numba supports, not just simple scalars.  So it will be possible to make things like `List[List[int]]` or `List[Dict[int, str]]`.

**Comment by mehrdadn** [comment]
Oh whoa I see, I didn't realize that. I think the approach overall makes sense then. The (more minor) concerns I have are that (a) there's no chance for the user to opt-in to the current default behavior if that's really what they wanted all along, and (b) explicit conversion by users might be slower than automatic conversion for smaller containers. They're not big deals but given that the current behavior still has its uses it'd be nice if it could be opted-in somehow.

**Comment by esc** [comment]
Another useful feature of the typed list will be, that it can grow as you append items to it, so you will not need to know ahead of time, how big it should be.

**Comment by ehsantn** [comment]
> > But you are right that this type of non-mutating list usage should keep working.
> 
> After discussing with the core devs, I recalled why even non-mutating list is going to be a challenge.
> 
> * When a Python list is used as argument to a jit'ed function, every item it holds must be scanned to determine the type.  This leads to an unexpected performance hit for the users.  If it's a list of heterogeneous types, the types will be coerced, which is the current behavior but also unexpected for users.
> * Mutation is hard to detect for the general case.  For a non-nested list of immutable types (i.e. ints, floats), the only mutation is by direct mutation on the list.  That is the simple case.  (Maybe we can do something special for this case.)  For others, there are indirect mutations, such as self-mutating methods in element of the list.

A common pattern I have seen is small read-only lists of scalars passed to Numba functions. For example, a list of parameters for computation, list of file names, etc. Unboxing the list into typed list is not performance critical. Can we support this pattern?

**Comment by NicolasHug** [comment]
I'm in this situation where I have a function (`_sum()`) that must support both arrays and lists, seamlessly. There is no reflection, I'm just reading what's in the array/list.

The `array_or_list` is something that users will provide, and it may be big: I don't want to convert it to an array or a typed list. 

What is the recommended solution here?  With the reflected lists deprecation, are we loosing the ability to deal with arrays and lists in a unified manner?

```py
from numba import njit
import numpy as np


@njit
def _sum(array_or_list):
    out = 0
    for i in range(len(array_or_list)):
        out += array_or_list[i]
    return out


def work(array_or_list):
    # do stuff

    s = _sum(array_or_list)

    # do other stuff

work(np.array([1, 2]))
work([1, 2])  # throws pending dep warning
```

**Comment by stuartarchibald** [comment]
In the current, as of Numba 0.45, situation, if your user passes a standard Python list to the `_sum` function Numba will have to unbox the list before use in the jitted code. Unboxing is the terminology used to describe creating a Numba internal list representation and then converting each element of the Python list into a native value and put that into the internal list representation. In the future situation (which you can also do now, electively, with `numba.typed.List`), there is no automatic conversion from a standard Python list to a internal list, instead a user has to declare the list type and build it for use in e.g. `_sum`, i.e. do the unboxing process by virtue of using the new `typed.List`. With time I anticipate convenience methods to help do this will be added.

One way right now of handling this might be:
```python
from numba import njit, typeof
from numba.typed import List
import numpy as np


@njit
def _sum(array_or_list):
    out = 0
    for i in range(len(array_or_list)):
        out += array_or_list[i]
    return out

def fix_array_or_list(l, dtype=None):
    if isinstance(l, np.ndarray):
        return l

    if not l:
        raise ValueError("List must contain at least one element")
    if dtype is None:
        dt = typeof(l[0])
    else:
        dt = dtype
    new_list = List.empty_list(dt)
    for x in l:
        new_list.append(x)
    return new_list

def work(array_or_list):
    # do stuff
    new_array_or_list = fix_array_or_list(array_or_list)
    s = _sum(new_array_or_list)

    # do other stuff

work(np.array([1, 2]))
work([1, 2])
```

**Comment by seibert** [comment]
After some feedback and thought on this, we are leaning toward solving this by have an "immutable typed list" and Python list arguments will be cast to that form.  If the list is modified, a compilation error would be raised.  The fundamental problem with reflected lists is the mutation, the performance issue with unboxing a Python list is a secondary issue.

**Comment by NicolasHug** [comment]
Thank you both for the insightful feedback.

Would the new "immutable typed list" also involve an `O(n)` unboxing?

I wanted to avoid converting to a numpy array specifically to avoid copying / unboxing, but if unboxing always happens then I might just always convert to an array (with the assumption that `np.asarray()` is as fast as numba's unboxing?)

**Comment by seibert** [comment]
Yes, it would involve `O(n)` unboxing, just as reflected lists do today.

**Comment by stuartarchibald** [comment]
Closing this issue as it seems to be resolved. If this is not the case please re-open with a comment about any item that appears to be unresolved. Many thanks.
